ctmap: add support for GC of DSR orphaned entries

jibi opened the pull request:
Let's start from some context: normally, when traffic on a node needs to
get SNAT-ed by the datapath, the following map entries get created:

CT table:

  | Dir | Source      | Destination |
  | --- | ----------- | ----------- |
  | OUT | src IP/port | dst IP/port |

NAT table:

  | Dir | Source      | Destination | Xlate target          |
  | --- | ----------- | ----------- | --------------------- |
  | OUT | src IP/port | dst IP/port | XLATE_SRC nat IP/port |
  | IN  | dst IP/port | nat IP/port | XLATE_DST src IP/port |

Given this, the current algorithm to clear orphan NAT entries (i.e.
entries that are not backed anymore by a related CT one) can be
described as follow:

for each ingress NAT entry:
  - derive the related egress CT key:

        source      = ingress NAT entry xlate target
        destination = ingress NAT entry source

  - if there's no entry for the egress CT key
    - derive the related egress NAT entry (aka original tuple/ingress
      reverse tuple):

          source       = ingress NAT entry xlate targe
          destination  = ingress NAT entry source
          xlate target = ingress NAT entry destination

    - delete both ingress and egress NAT entries

This commit is about adding support for collecting orphaned NAT entries
created by the DSR loadbalancer mode.

In DSR mode, when a client connects to a NodePort service, the traffic
flow will be something like:

- client connects to a second node where the NodePort service is running
- this second node selects a backend and (unless the backend is running
  locally) forwards the traffic to a third node where the backend is
  running, while:
  - keeping the original client source IP
  - encoding the original destination in an IP option
  - rewriting the destination to the backend IP
- in the third node, the original destination is extracted from the IP
  option and the following bpf entries are created:

CT table:

  | Dir | Source      | Destination | Flags |
  | --- | ----------- | ----------- | ----- |
  | IN  | src IP/port | dst IP/port | DSR   |

NAT:

  | Dir | Source      | Destination | Xlate target     |
  | --- | ----------- | ----------- | ---------------- |
  | OUT | dst IP/port | src IP/port | XLATE_SRC nat IP |

  where:

  - src IP is the IP of the client
  - dst IP is the IP of the backend running on the node
  - nat IP is the IP of the nodeport node (second node), which was
    encoded by the node itself in an IP option before forwarding traffic
    to the third node.
    This is used to implement DSR as the reply traffic leaving the
    current node is forwarded back to the original client with this IP
    as source

Considering the entries that get created by the regular (i.e. non DSR)
NAT operations, to implement GC of the DSR SNAT entries we can extend
the algorithm as follows:

for each egress NAT entry:
  - derive the related ingress CT key:

        source      = ingress NAT entry destination
        destination = ingress NAT entry source

  - derive the related egress CT key:

        source      = ingress NAT entry source
        destination = ingress NAT entry destination

  - if both the ingress and egress CT keys don't have a corresponding
    entry:
    - delete the egress NAT entry

If both ingress and egress CT keys for a given NAT entry don't exist we
can assume the entry is orphan, as the lack of the ingress key will
cover the DSR case while the lack of the egress key will cover the
"regular" one.

There's one last catch: to simplify this writing, we discussed the
_logical_ content of the CT keys, but in practice every key has the
source and destination addresses swapped because of the way the datapath
creates these entries.

Fixes: https://github.com/cilium/cilium/issues/21346

jibi committed:
ctmap: add support for GC of DSR orphaned entries

Let's start from some context: normally, when traffic on a node needs to
get SNAT-ed by the datapath, the following map entries get created:

CT table:

  | Dir | Source      | Destination |
  | --- | ----------- | ----------- |
  | OUT | src IP/port | dst IP/port |

NAT table:

  | Dir | Source      | Destination | Xlate target          |
  | --- | ----------- | ----------- | --------------------- |
  | OUT | src IP/port | dst IP/port | XLATE_SRC nat IP/port |
  | IN  | dst IP/port | nat IP/port | XLATE_DST src IP/port |

Given this, the current algorithm to clear orphan NAT entries (i.e.
entries that are not backed anymore by a related CT one) can be
described as follow:

for each ingress NAT entry:
  - derive the related egress CT key:

        source      = ingress NAT entry xlate target
        destination = ingress NAT entry source

  - if there's no entry for the egress CT key
    - derive the related egress NAT entry (aka original tuple/ingress
      reverse tuple):

          source       = ingress NAT entry xlate targe
          destination  = ingress NAT entry source
          xlate target = ingress NAT entry destination

    - delete both ingress and egress NAT entries

This commit is about adding support for collecting orphaned NAT entries
created by the DSR loadbalancer mode.

In DSR mode, when a client connects to a NodePort service, the traffic
flow will be something like:

- client connects to a second node where the NodePort service is running
- this second node selects a backend and (unless the backend is running
  locally) forwards the traffic to a third node where the backend is
  running, while:
  - keeping the original client source IP
  - encoding the original destination in an IP option
  - rewriting the destination to the backend IP
- in the third node, the original destination is extracted from the IP
  option and the following bpf entries are created:

CT table:

  | Dir | Source      | Destination | Flags |
  | --- | ----------- | ----------- | ----- |
  | IN  | src IP/port | dst IP/port | DSR   |

NAT:

  | Dir | Source      | Destination | Xlate target     |
  | --- | ----------- | ----------- | ---------------- |
  | OUT | dst IP/port | src IP/port | XLATE_SRC nat IP |

  where:

  - src IP is the IP of the client
  - dst IP is the IP of the backend running on the node
  - nat IP is the IP of the nodeport node (second node), which was
    encoded by the node itself in an IP option before forwarding traffic
    to the third node.
    This is used to implement DSR as the reply traffic leaving the
    current node is forwarded back to the original client with this IP
    as source

Considering the entries that get created by the regular (i.e. non DSR)
NAT operations, to implement GC of the DSR SNAT entries we can extend
the algorithm as follows:

for each egress NAT entry:
  - derive the related ingress CT key:

        source      = ingress NAT entry destination
        destination = ingress NAT entry source

  - derive the related egress CT key:

        source      = ingress NAT entry source
        destination = ingress NAT entry destination

  - if both the ingress and egress CT keys don't have a corresponding
    entry:
    - delete the egress NAT entry

If both ingress and egress CT keys for a given NAT entry don't exist we
can assume the entry is orphan, as the lack of the ingress key will
cover the DSR case while the lack of the egress key will cover the
"regular" one.

There's one last catch: to simplify this writing, we discussed the
_logical_ content of the CT keys, but in practice every key has the
source and destination addresses swapped because of the way the datapath
creates these entries.

Fixes: #21346
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
tuple: add SwapAddresses() helpers

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

julianwiedmann commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -627,6 +620,21 @@ func PurgeOrphanNATEntries(ctMapTCP, ctMapAny *Map) *NatGCStats {
 				}
 			} else {
 				stats.IngressAlive += 1
+				stats.EgressAlive += 1
```
This is because we assume that such ingress NAT entry always has a related egress NAT entry, correct? Why can't we handle this in the case below, once we visit that egress NAT entry?

jibi replied:
Right, we are already accounting for this in the case below, I should remove it :+1:

julianwiedmann commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -627,6 +620,21 @@ func PurgeOrphanNATEntries(ctMapTCP, ctMapAny *Map) *NatGCStats {
+				stats.EgressAlive += 1
+			}
+		} else if natKey.GetFlags()&tuple.TUPLE_F_OUT == tuple.TUPLE_F_OUT {
+			ingressNatKey := ingressNatKeyFromEgressNatKeyVal(natKey, natVal)
```
For my understanding - instead of checking for "no Ingress CT entry AND no Ingress NAT entry", we could also check for "no Ingress CT entry AND no Egress CT entry". Correct?

Should we raise some warning if we find an Ingress CT entry, but the entry doesn't have the DSR flag?

jibi replied:
>For my understanding - instead of checking for "no Ingress CT entry AND no Ingress NAT entry", we could also check for "no Ingress CT entry AND no Egress CT entry". Correct?

Yep, that is correct also to my understanding :+1: 

I decided to go for the CT and NAT entries check to make it more explicit (i.e. it's a DSR SNAT entry and it doesn't have a related NAT entry), but perhaps once we are happy with this we can evaluate the incress/egress CT check and maybe unify the 2 cases (just food for thoughts)

> Should we raise some warning if we find an Ingress CT entry, but the entry doesn't have the DSR flag?

I'm not sure here, I would imagine it's possible to have ingress CT entries without NAT being involved

julianwiedmann replied:
> > For my understanding - instead of checking for "no Ingress CT entry AND no Ingress NAT entry", we could also check for "no Ingress CT entry AND no Egress CT entry". Correct?
> 
> Yep, that is correct also to my understanding +1
> 
> I decided to go for the CT and NAT entries check to make it more explicit (i.e. it's a DSR SNAT entry and it doesn't have a related NAT entry), 

yeah, that's what I figured :). I understand where you're coming from (focusing it exactly on how the situation looks like for DSR). But I feel the reasoning "if there's no ingress CT and no egress CT, then obviously this NAT entry is stale" is more straight-forward, and requires less understanding of how NAT/CT works for DSR in detail. Maybe that tips the balance, both approaches being equal? Either way is alright.


> > Should we raise some warning if we find an Ingress CT entry, but the entry doesn't have the DSR flag?
> 
> I'm not sure here, I would imagine it's possible to have ingress CT entries without NAT being involved

Ah sorry. I specifically meant for the "we are dealing with a DSR entry" path. Here it would be nice to know if we unexpectedly match something that's **not** DSR.

jibi replied:
> I feel the reasoning "if there's no ingress CT and no egress CT, then obviously this NAT entry is stale" is more straight-forward, and requires less understanding of how NAT/CT works for DSR in detail

Fair enough, my initial aim was to narrow down the changes to the DSR case in order to avoid introducing regressions in the logic already in place, but now that we have a clearer picture of what's going on with these entries I'm mostly convinced that approach should be fine as well, so I'll give it a try :+1: 

> Ah sorry. I specifically meant for the "we are dealing with a DSR entry" path. Here it would be nice to know if we unexpectedly match something that's not DSR.

Got it (but switching to checking only CT entries means we can't detect this case anymore)

julianwiedmann reviewed (commented):
Thanks @jibi! This all sounds sane, but some questions to make sure my understanding is correct ...

jibi committed:
maps/ctmap: switch utils.go to SwapAddresses() tuples' helpers

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

julianwiedmann commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -628,6 +621,19 @@ func PurgeOrphanNATEntries(ctMapTCP, ctMapAny *Map) *NatGCStats {
+					if deleted, _ := natMap.Delete(natKey); deleted {
+						stats.EgressDeleted += 1
+					}
+				}
```
For the inner `if` condition we're missing an
```
} else {
	stats.EgressAlive += 1
}
```

But this is fixed up by the last patch in the series. Might still want to get it right here, so that we can revert without regression.

jibi replied:
Good shout, updated :+1:

julianwiedmann reviewed (approved):
Thanks! Good call on doing the refactor at the end :+1:.

Just one minor thing that's worth addressing imho. But either way lgtm.

jibi committed:
ctmap: refactor NAT GC logic

After adding support to the GC for NAT entries created by the DSR logic,
we ended up handling twice the case for orphan egress entries as
currently when we find an orphan ingress one we derive the related
egress one and delete it as well.

Since this case is already handled by the logic that goes through
egress NAT entries, simply remove it.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>
